Hypothesis fails with ForwardRefs and recursive pydantic models

I would like to use hypothesis to generate instances of pydantic models, but it looks like it cannot handle recursive models with ForwardRefs.

(Quite) minimal example triggering the problem:

```python
from pydantic import BaseModel
from typing import Optional, Union
from hypothesis import given, strategies as st

class Dummy(BaseModel):
    x: Optional[Union[int, Dummy]]

Dummy.update_forward_refs()


@given(st.builds(Dummy))
def test_dummy(obj):
    print(obj.json(indent=2))
```

When running this with pytest, it fails with:
```
E           hypothesis.errors.ResolutionFailed: Could not resolve ForwardRef('Optional[Union[int, Dummy]]') to a strategy; consider using register_type_strategy

.../site-packages/hypothesis/strategies/_internal/types.py:442: ResolutionFailed
```

**Extra info:**
* Python 3.8
* pydantic 1.10.2
* hypothesis 6.58
* pytest 7.0.1

--- a/hypothesis-python/RELEASE.rst
+++ b/hypothesis-python/RELEASE.rst
@@ -1,0 +1,6 @@
+RELEASE_TYPE: minor
+
+This release improves Hypothesis' ability to resolve forward references in 
+type annotations. It fixes a bug that prevented 
+:func:`~hypothesis.strategies.builds` from being used with `pydantic models that
+possess updated forward references <https://pydantic-docs.helpmanual.io/usage/postponed_annotations/>`__. See :issue:`3519`.

--- a/hypothesis-python/src/hypothesis/internal/compat.py
+++ b/hypothesis-python/src/hypothesis/internal/compat.py
@@ -111,8 +111,13 @@ def get_type_hints(thing):
                     and is_a_type(p.annotation)
                     and p.annotation is not p.empty
                 ):
+                    if isinstance(p.annotation, typing.ForwardRef) and not isinstance(
+                        hints[p.name], typing.ForwardRef
+                    ):
+                        continue
+
                     if p.default is None:
-                        hints[p.name] = typing.Optional[p.annotation]
+                        hints[p.name] = typing.Optional[p.annotation]  # type: ignore
                     else:
                         hints[p.name] = p.annotation
     except (AttributeError, TypeError, NameError):  # pragma: no cover

--- a/hypothesis-python/src/hypothesis/internal/compat.py
+++ b/hypothesis-python/src/hypothesis/internal/compat.py
@@ -13,7 +13,7 @@ import inspect
 import platform
 import sys
 import typing
-from typing import Any, ForwardRef, List, Tuple, Union
+from typing import Any, ForwardRef, Tuple
 
 try:
     from typing import get_args
@@ -23,7 +23,7 @@ except ImportError:
 
     def get_args(
         tp: Any,
-    ) -> Union[Tuple[type, ...], Tuple[List[type], type]]:  # pragma: no cover
+    ) -> Tuple[Any, ...]:  # pragma: no cover
         """
         Examples
         --------
@@ -155,13 +155,14 @@ def get_type_hints(thing):
                         isinstance(sig_hint, ForwardRef)
                         and not isinstance(hint, ForwardRef)
                         for sig_hint, hint in zip(
-                            _hint_and_args(p.annotation), _hint_and_args(hints[p.name])
+                            _hint_and_args(p.annotation),
+                            _hint_and_args(hints.get(p.name, Any)),
                         )
                     ):
                         p_hint = hints[p.name]
-
+                    print(p_hint, p.default)
                     if p.default is None:
-                        hints[p.name] = typing.Optional[p_hint]  # type: ignore
+                        hints[p.name] = typing.Optional[p_hint]
                     else:
                         hints[p.name] = p_hint
     except (AttributeError, TypeError, NameError):  # pragma: no cover

--- a/hypothesis-python/src/hypothesis/internal/compat.py
+++ b/hypothesis-python/src/hypothesis/internal/compat.py
@@ -160,7 +160,6 @@ def get_type_hints(thing):
                         )
                     ):
                         p_hint = hints[p.name]
-                    print(p_hint, p.default)
                     if p.default is None:
                         hints[p.name] = typing.Optional[p_hint]
                     else:

--- a/hypothesis-python/src/hypothesis/internal/compat.py
+++ b/hypothesis-python/src/hypothesis/internal/compat.py
@@ -150,7 +150,7 @@ def get_type_hints(thing):
                     p_hint = p.annotation
 
                     # Defer to `get_type_hints` if signature annotation is, or
-                    # contains, a forward refence that is otherwise resolved.
+                    # contains, a forward reference that is otherwise resolved.
                     if any(
                         isinstance(sig_hint, ForwardRef)
                         and not isinstance(hint, ForwardRef)

--- a/hypothesis-python/src/hypothesis/internal/compat.py
+++ b/hypothesis-python/src/hypothesis/internal/compat.py
@@ -13,6 +13,37 @@ import inspect
 import platform
 import sys
 import typing
+from typing import Any, ForwardRef, List, Tuple, Union
+
+try:
+    from typing import get_args
+except ImportError:  # pragma: no cover
+    # remove at Python 3.7 end-of-life
+    from collections.abc import Callable as _Callable
+
+    def get_args(
+        tp: Any,
+    ) -> Union[Tuple[type, ...], Tuple[List[type], type]]:  # pragma: no cover
+        """
+        Examples
+        --------
+        >>> assert get_args(int) == ()
+        >>> assert get_args(Dict[str, int]) == (str, int)
+        >>> assert get_args(Union[int, Union[T, int], str][int]) == (int, str)
+        >>> assert get_args(Union[int, Tuple[T, int]][str]) == (int, Tuple[str, int])
+        >>> assert get_args(Callable[[], T][int]) == ([], int)
+        """
+        if hasattr(tp, "__origin__") and hasattr(tp, "__args__"):
+            args = tp.__args__
+            if (
+                getattr(tp, "__origin__", None) is _Callable
+                and args
+                and args[0] is not Ellipsis
+            ):
+                args = (list(args[:-1]), args[-1])
+            return args
+        return ()
+
 
 try:
     BaseExceptionGroup = BaseExceptionGroup
@@ -71,6 +102,10 @@ def is_typed_named_tuple(cls):
     )
 
 
+def _hint_and_args(x):
+    return (x,) + get_args(x)
+
+
 def get_type_hints(thing):
     """Like the typing version, but tries harder and never errors.
 
@@ -114,8 +149,14 @@ def get_type_hints(thing):
 
                     p_hint = p.annotation
 
-                    if isinstance(p.annotation, typing.ForwardRef) and not isinstance(
-                        hints[p.name], typing.ForwardRef
+                    # Defer to `get_type_hints` if signature annotation is, or
+                    # contains, a forward refence that is otherwise resolved.
+                    if any(
+                        isinstance(sig_hint, ForwardRef)
+                        and not isinstance(hint, ForwardRef)
+                        for sig_hint, hint in zip(
+                            _hint_and_args(p.annotation), _hint_and_args(hints[p.name])
+                        )
                     ):
                         p_hint = hints[p.name]
 

--- a/hypothesis-python/src/hypothesis/internal/compat.py
+++ b/hypothesis-python/src/hypothesis/internal/compat.py
@@ -111,15 +111,18 @@ def get_type_hints(thing):
                     and is_a_type(p.annotation)
                     and p.annotation is not p.empty
                 ):
+
+                    p_hint = p.annotation
+
                     if isinstance(p.annotation, typing.ForwardRef) and not isinstance(
                         hints[p.name], typing.ForwardRef
                     ):
-                        continue
+                        p_hint = hints[p.name]
 
                     if p.default is None:
-                        hints[p.name] = typing.Optional[p.annotation]  # type: ignore
+                        hints[p.name] = typing.Optional[p_hint]  # type: ignore
                     else:
-                        hints[p.name] = p.annotation
+                        hints[p.name] = p_hint
     except (AttributeError, TypeError, NameError):  # pragma: no cover
         pass
 

--- a/hypothesis-python/src/hypothesis/internal/compat.py
+++ b/hypothesis-python/src/hypothesis/internal/compat.py
@@ -17,7 +17,7 @@ from typing import Any, ForwardRef, List, Tuple, Union
 
 try:
     from typing import get_args
-except ImportError:  # pragma: no cover
+except ImportError:
     # remove at Python 3.7 end-of-life
     from collections.abc import Callable as _Callable
 

--- a/hypothesis-python/tests/cover/test_compat.py
+++ b/hypothesis-python/tests/cover/test_compat.py
@@ -11,7 +11,7 @@
 import math
 from dataclasses import dataclass
 from inspect import Parameter, Signature, signature
-from typing import ForwardRef
+from typing import ForwardRef, Optional
 
 import pytest
 

--- a/hypothesis-python/tests/cover/test_compat.py
+++ b/hypothesis-python/tests/cover/test_compat.py
@@ -51,7 +51,7 @@ def test_no_type_hints():
 
 @dataclass
 class Foo:
-    x: "Foo"
+    x: "Foo" = None  # type: ignore
 
 
 Foo.__signature__ = signature(Foo).replace(  # type: ignore
@@ -67,4 +67,4 @@ Foo.__signature__ = signature(Foo).replace(  # type: ignore
 
 def test_resolve_fwd_refs():
     # See: https://github.com/HypothesisWorks/hypothesis/issues/3519
-    assert get_type_hints(Foo)["x"] is Foo
+    assert get_type_hints(Foo)["x"] is Optional[Foo]

--- a/hypothesis-python/tests/cover/test_compat.py
+++ b/hypothesis-python/tests/cover/test_compat.py
@@ -60,6 +60,7 @@ Foo.__signature__ = signature(Foo).replace(  # type: ignore
             "x",
             Parameter.POSITIONAL_OR_KEYWORD,
             annotation=ForwardRef("Foo"),
+            default=None,
         )
     ]
 )

--- a/hypothesis-python/tests/cover/test_compat.py
+++ b/hypothesis-python/tests/cover/test_compat.py
@@ -11,7 +11,7 @@
 import math
 from dataclasses import dataclass
 from inspect import Parameter, Signature, signature
-from typing import ForwardRef, Optional
+from typing import ForwardRef, Optional, Union
 
 import pytest
 
@@ -65,6 +65,25 @@ Foo.__signature__ = signature(Foo).replace(  # type: ignore
 )
 
 
-def test_resolve_fwd_refs():
+@dataclass
+class Bar:
+    x: Optional[Union[int, "Bar"]]
+
+
+Bar.__signature__ = signature(Bar).replace(  # type: ignore
+    parameters=[
+        Parameter(
+            "x",
+            Parameter.POSITIONAL_OR_KEYWORD,
+            annotation=Optional[Union[int, ForwardRef("Bar")]],  # type: ignore
+        )
+    ]
+)
+
+
+@pytest.mark.parametrize(
+    "obj,expected", [(Foo, Optional[Foo]), (Bar, Optional[Union[int, Bar]])]
+)
+def test_resolve_fwd_refs(obj, expected):
     # See: https://github.com/HypothesisWorks/hypothesis/issues/3519
-    assert get_type_hints(Foo)["x"] is Optional[Foo]
+    assert get_type_hints(obj)["x"] == expected

--- a/hypothesis-python/tests/cover/test_compat.py
+++ b/hypothesis-python/tests/cover/test_compat.py
@@ -9,7 +9,9 @@
 # obtain one at https://mozilla.org/MPL/2.0/.
 
 import math
-from inspect import Parameter, Signature
+from dataclasses import dataclass
+from inspect import Parameter, Signature, signature
+from typing import ForwardRef
 
 import pytest
 
@@ -45,3 +47,24 @@ class WeirdSig:
 
 def test_no_type_hints():
     assert get_type_hints(WeirdSig) == {}
+
+
+@dataclass
+class Foo:
+    x: "Foo"
+
+
+Foo.__signature__ = signature(Foo).replace(  # type: ignore
+    parameters=[
+        Parameter(
+            "x",
+            Parameter.POSITIONAL_OR_KEYWORD,
+            annotation=ForwardRef("Foo"),
+        )
+    ]
+)
+
+
+def test_resolve_fwd_refs():
+    # See: https://github.com/HypothesisWorks/hypothesis/issues/3519
+    assert get_type_hints(Foo)["x"] is Foo

